### PR TITLE
Benchmark job with backoff limit per index

### PR DESCRIPTION
**Pull request by mimowo**
#### What type of PR is this?

/kind cleanup

#### Which issue(s) this PR fixes:

Part of: https://github.com/kubernetes/enhancements/issues/3850

#### Special notes for your reviewer:

1. Experiments on my local machine running the Benchmark integration test for jobs *with* failures using the `BenchmarkLargeFailureHandling` test. Here we have  
```
b.Cleanup(setDurationDuringTest(&jobcontroller.DefaultJobPodFailureBackOff, fastPodFailureBackoff))
b.Cleanup(setDurationDuringTest(&jobcontroller.MaxJobPodFailureBackOff, fastPodFailureBackoff))
```
to minimize the timing differences due to different backoff strategies.

The results of the benchmark are put the KEP update PR: https://github.com/kubernetes/enhancements/pull/4321. 

Also, lowered the nPods parameter to 10 and 100 to avoid issues with long running time on CI. The parameters can be easily increased for local testing.

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
https://github.com/kubernetes/enhancements/tree/master/keps/sig-apps/3850-backoff-limits-per-index-for-indexed-jobs
```

### CODE REVIEW

**Comment by mimowo** [comment]
@alculquicondor PTAL at the benchmark results. 
1. Do you think this warrants further investigation at this point?
2. Do you think we want something like this merged, or the one time experiment on-demand is enough: https://github.com/kubernetes/kubernetes/pull/121393? I'm a little bit worried about merging the benchmark given our issues with the runtime for integration tests.

**Review by alculquicondor — COMMENTED** [review]
Why is this benchmark relevant?
If there are no Pod failures, why would having `backoffLimitPerIndex` have an impact?

**Comment by mimowo** [comment]
> Why is this benchmark relevant?
> If there are no Pod failures, why would having `backoffLimitPerIndex` have an impact?

This is a sanity benchmark to make sure there is no significant penalty for stepping into some new branches of code like [here](https://github.com/kubernetes/kubernetes/blob/8453eb0c24ea61794157b451e4442ea570abce45/pkg/controller/job/job_controller.go#L846) or [here](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/job/job_controller.go#L1526-L1531). 

However, to verify the new branches don't add significant penalty at scale we would need to have many failures, probably >1000. At this point, delay due to expotential backoff would take over. Since the delay patterns are different between global and perIndex backoffLimit I'm not sure how to design the proper full benchmark.

**Comment by mimowo** [comment]
The results show virtually no difference;
* Regular for N=1000: 21.125965065s
* BackoffLimitPerIndex for N=1000: 21.134007616s
* Regular for N=10_000: 3m22.003681845s
* BackoffLimitPerIndex for N=10_000: 3m22.070514111s

Measuring using b.Elapsed(), previous experiments where reported for the entire test, from start to end, so it could also explain higher varation.

**Comment by mimowo** [comment]
I have repeated the results for the test without failures to use the same methodology of measuring (based on b.Elapsed(), not the whole test). Now it shows even less of a difference.

**Comment by mimowo** [comment]
@alculquicondor PTAL at the revised results. 

Do you think the new test with failures should make into master? I'm worried about adding integration tests which operate on >=1000 pods. Even the currently existing test is a stretch IMO. 

Maybe we could add the new test if we lowered scale for the current one, say to 10 & 100. Then, if anyone is interested then can bump the numbers locally.

File: pkg/controller/controller_utils.go [hunk 1]
```diff
@@ -952,7 +952,7 @@ func FilterActivePods(logger klog.Logger, pods []*v1.Pod) []*v1.Pod {
 		if IsPodActive(p) {
 			result = append(result, p)
 		} else {
-			logger.V(4).Info("Ignoring inactive pod", "pod", klog.KObj(p), "phase", p.Status.Phase, "deletionTime", p.DeletionTimestamp)
```
Comment (alculquicondor) [inline pkg/controller/controller_utils.go:955]: leave this for a separate PR

File: staging/src/k8s.io/apimachinery/pkg/util/wait/wait.go [hunk 2]
```diff
@@ -34,7 +34,7 @@ import (
 // The value for timeout should effectively be "forever." Obviously we don't want our tests to truly lock up forever, but 30s
 // is long enough that it is effectively forever for the things that can slow down a run on a heavily contended machine
 // (GC, seeks, etc), but not so long as to make a developer ctrl-c a test run if they do happen to break that test.
-var ForeverTestTimeout = time.Second * 30
+var ForeverTestTimeout = time.Minute * 5
```
Comment (alculquicondor) [inline staging/src/k8s.io/apimachinery/pkg/util/wait/wait.go:37]: I don't think this is a good idea. If you need this for a specific test, do it there.

File: test/integration/job/job_test.go [hunk 3]
```diff
@@ -2117,6 +2139,140 @@ func BenchmarkLargeIndexedJob(b *testing.B) {
 				}
 				b.StartTimer()
 			}
+			b.Logf("Time elapsed for %s: %v", name, b.Elapsed())
```
Comment (alculquicondor) [inline test/integration/job/job_test.go:2142]:
```suggestion
			b.Logf("Time elapsed: %v", name, b.Elapsed())
```

The name should already be included by the framework.

File: test/integration/job/job_test.go [hunk 4]
```diff
@@ -2117,6 +2139,140 @@ func BenchmarkLargeIndexedJob(b *testing.B) {
...
+							Limit: 1000,
+						})
+					if err != nil {
+						b.Fatalf("Failed to cleanup Pods: %v", err)
+					}
+				}
+				err = clientSet.BatchV1().Jobs(jobObj.Namespace).Delete(ctx, jobObj.Name, metav1.DeleteOptions{})
+				if err != nil {
+					b.Fatalf("Failed to cleanup Job: %v", err)
+				}
```
Comment (alculquicondor) [inline test/integration/job/job_test.go:2272]: Let's put this in a function for both test to use and call with `b.CleanUp`

File: test/integration/job/job_test.go [hunk 5]
```diff
@@ -2117,6 +2139,140 @@ func BenchmarkLargeIndexedJob(b *testing.B) {
...
+					return true, nil
+				}); err != nil {
+					b.Fatalf("Could not succeed the remaining %d pods: %v", remaining, err)
+				}
+				validateJobPodsStatus(ctx, b, clientSet, jobObj, podsByStatus{
+					Ready:     ptr.To[int32](0),
+					Succeeded: int(tc.nPods),
+					Failed:    int(tc.nPods),
+				})
+				validateJobSucceeded(ctx, b, clientSet, jobObj)
```
Comment (alculquicondor) [inline test/integration/job/job_test.go:2248]: why do we need to make the job succeed as part of the cleanup?

**Review by alculquicondor — COMMENTED** [review]
I think it's valuable to keep this for a bit.
You should probably copy some of the results to the KEP, for future reference when considering the graduation to stable.

File: test/integration/job/job_test.go (hunk 5 above)
Comment (mimowo) [reply test/integration/job/job_test.go:2248]: I don't think we need, we could make the test shorter by this. It was sort of sanity test for me to verify the test is doing want. It is not measured.

File: pkg/controller/controller_utils.go (hunk 1 above)
Comment (mimowo) [reply pkg/controller/controller_utils.go:955]: Just wanted to make sure the constant panics don't impact the benchmarks.

File: test/integration/job/job_test.go (hunk 5 above)
Comment (mimowo) [reply test/integration/job/job_test.go]: Extracted

File: test/integration/job/job_test.go (hunk 4 above)
Comment (mimowo) [reply test/integration/job/job_test.go]: Done

File: test/integration/job/job_test.go (hunk 3 above)
Comment (mimowo) [reply test/integration/job/job_test.go]: Actually, this is not needed. The benchmark measurements are reported even without the log, I just missed the lines. Here is an example: 
```
BenchmarkLargeIndexedJob/regular_indexed_job_without_failures;_size=10-48                       	       1	3034342185 ns/op	14391160 B/op	  164352 allocs/op
BenchmarkLargeIndexedJob/regular_indexed_job_without_failures;_size=100-48                      	       1	3050613253 ns/op	111100464 B/op	 1324757 allocs/op
```

File: staging/src/k8s.io/apimachinery/pkg/util/wait/wait.go (hunk 2 above)
Comment (mimowo) [reply staging/src/k8s.io/apimachinery/pkg/util/wait/wait.go]: Removed, and introduced `customTimeout` only used for large N.

**Comment by mimowo** [comment]
>  I think it's valuable to keep this for a bit.
You should probably copy some of the results to the KEP, for future reference when considering the graduation to stable.

SGTM. I have prepared the KEP update: https://github.com/kubernetes/enhancements/pull/4321

File: pkg/controller/controller_utils.go (hunk 1 above)
Comment (mimowo) [reply pkg/controller/controller_utils.go]: Done

File: test/integration/job/job_test.go [hunk 6]
```diff
@@ -2127,38 +2155,135 @@ func BenchmarkLargeIndexedJob(b *testing.B) {
...
+					Terminating: ptr.To[int32](0),
+				}, timeout)
+				b.StopTimer()
 			}
 		})
 	}
 }
 
+// cleanUp deletes all pods and the job
+func cleanUp(ctx context.Context, b *testing.B, clientSet clientset.Interface, jobObj *batchv1.Job) error {
```
Comment (alculquicondor) [inline test/integration/job/job_test.go:2264]:
```suggestion
func cleanUp(ctx context.Context, clientSet clientset.Interface, jobObj *batchv1.Job) error {
```

**Review by alculquicondor — COMMENTED** [review]
/approve

**Comment by alculquicondor** [comment]
/retitle Benchmark job with backoff limit per index

**Comment by mimowo** [comment]
cc @kannon92 @pohly 
would you like to review as well?

**Comment by mimowo** [comment]
/test pull-kubernetes-unit 
retry as this is failure unrelated with this PR, but it might be good to look into this deeper as this is a job failure: https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/121393/pull-kubernetes-unit/1719392853026672640
FYI @alculquicondor @dejanzele

**Comment by mimowo** [comment]
> /test pull-kubernetes-unit retry as this is failure unrelated with this PR, but it might be good to look into this deeper as this is a job failure: https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/121393/pull-kubernetes-unit/1719392853026672640 FYI @alculquicondor @dejanzele

Ticketed: https://github.com/kubernetes/kubernetes/issues/121652

**Comment by alculquicondor** [comment]
oops
/lgtm

**Comment by pacoxu** [comment]
This is a test change. I suppose that we can add milestone v1.29 for this.

**Comment by mimowo** [comment]
Yes, I think it makes sense to add the milestone.
